tuna/gen_fastdb.py

```diff
@@ -0,0 +1,298 @@
+  if num_duplicates > 0:
+    if verbosity >= 1:
+      logging.log(f'{num_duplicates} duplicates found! generating report...',
+                  end_char='\r')
+      duplicates = finddb[finddb.duplicated(subset=cols_with_conv_params +
+                                            ['solver'],
+                                            keep=False)]
+      duplicates = duplicates.sort_values(cols_with_conv_params +
+                                          ['solver', 'kernel_time'])
+      duplicates = duplicates.groupby(cols_with_conv_params + ['solver'])
+
+      duplicates_str = as_heading("Duplicates") + '\n'
+      for i, (_, df) in enumerate(duplicates):  # pylint: disable=invalid-name
+        for _, row in df.iterrows():
+          conv_params_str = "+ " if i % 2 == 0 else "- "
+          for colname, val in zip(cols_with_conv_params,
+                                  row[cols_with_conv_params]):
+            conv_params_str += f'{colname} {val}, '
+          duplicates_str += f"{conv_params_str}\n  solver: {row['solver']},  " +\
+                    f"kernel_time: {row['kernel_time']}\n\n"
+
+      logging.log(duplicates_str, silent=True)
+
+    if strict:
+      raise AssertionError(
+          f'{num_duplicates} duplicate entries detected in {tag}finddb')
+    logging.warning(
+        f'{num_duplicates} duplicate entries delected in {tag}finddb')
+    issues.append(DUPLICATE_ENTRIES)
+
+  if not issues:
+    logging.success(f'{tag}finddb passed all checks!')
+
+  return issues
+
+
+def describe_fastdb(fastdb, tag=''):
+  """ describe fastdb (num. of entries in it, etc.) """
+  if len(fastdb) == 0:
+    logging.warning(f'{tag}fastdb empty!')
+  else:
+    logging.info(
+        f'Total entries in {tag}fastdb (=unique configs in {tag}finddb): %d' %
+        len(fastdb))
+    logging.info(f'Number of unique solvers in {tag}fastdb: %d' %
+                 len(fastdb['solver'].unique()))
+
+
+def load_fastdb(fastdb_pickle_filename):
+  """ load fastdb from pickle """
+  fastdb = io_tools.safe_load(None, fastdb_pickle_filename,
+                              df_tools.from_pickle)
+  describe_fastdb(fastdb)
+  return fastdb
+
+
+def finddb_to_nonthresholded_fastdb(finddb,
+                                    cols_with_conv_params=None,
+                                    n_fastest=1,
+                                    tag=''):
+  """ generate fastdb from finddb without thresholding/removing out any solvers """
+  if cols_with_conv_params is None:
+    cols_with_conv_params = _DEFAULT_COLS_WITH_CONV_PARAMS
+
+  num_duplicates = finddb.duplicated(subset=cols_with_conv_params +
+                                     ['solver']).sum()
+  if num_duplicates > 0:
+    logging.warning(f'resolved the {num_duplicates} duplicate (fdb_key, solver)' +\
+            f' pairs found in {tag}finddb by keeping the entries with fastest solver')
+  sorted_finddb = finddb.sort_values(cols_with_conv_params + ['kernel_time'])
```

Comment:
this sorting can and should be done in mysql, by using the func max() attribute in SQLAlchemy. Or, if you want to select all the entries including duplicates, you can sort by kernel_time asc or desc and then you can pick the fastest entry in constant time in python

Reply:
Don't really know MySQL, so I just converted conv_find_db table to a pandas dataframe in `gen_finddb.py`, so that all the scripts from there on process data frames